src/storage/azure/storage.c

```diff
@@ -0,0 +1,729 @@
+        FUNCTION_LOG_PARAM(BOOL, param.errorOnMissing);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(file != NULL);
+    ASSERT(!param.errorOnMissing);
+
+    storageAzureRequestP(this, HTTP_VERB_DELETE_STR, file, .allowMissing = true);
+
+    FUNCTION_LOG_RETURN_VOID();
+}
+
+/**********************************************************************************************************************************/
+static const StorageInterface storageInterfaceAzure =
+{
+    .info = storageAzureInfo,
+    .infoList = storageAzureInfoList,
+    .newRead = storageAzureNewRead,
+    .newWrite = storageAzureNewWrite,
+    .pathRemove = storageAzurePathRemove,
+    .remove = storageAzureRemove,
+};
+
+Storage *
+storageAzureNew(
+    const String *path, bool write, StoragePathExpressionCallback pathExpressionFunction, const String *container,
+    const String *account, const String *key, size_t blockSize, const String *host, unsigned int port, TimeMSec timeout,
+    bool verifyPeer, const String *caFile, const String *caPath)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(STRING, path);
+        FUNCTION_LOG_PARAM(BOOL, write);
+        FUNCTION_LOG_PARAM(FUNCTIONP, pathExpressionFunction);
+        FUNCTION_LOG_PARAM(STRING, container);
+        FUNCTION_TEST_PARAM(STRING, account);
+        FUNCTION_TEST_PARAM(STRING, key);
+        FUNCTION_LOG_PARAM(SIZE, blockSize);
+        FUNCTION_LOG_PARAM(STRING, host);
+        FUNCTION_LOG_PARAM(UINT, port);
+        FUNCTION_LOG_PARAM(TIME_MSEC, timeout);
+        FUNCTION_LOG_PARAM(BOOL, verifyPeer);
+        FUNCTION_LOG_PARAM(STRING, caFile);
+        FUNCTION_LOG_PARAM(STRING, caPath);
+    FUNCTION_LOG_END();
+
+    ASSERT(path != NULL);
+
+    Storage *this = NULL;
+
+    MEM_CONTEXT_NEW_BEGIN("StorageAzure")
+    {
+        StorageAzure *driver = memNew(sizeof(StorageAzure));
+
+        *driver = (StorageAzure)
+        {
+            .memContext = MEM_CONTEXT_NEW(),
+            .interface = storageInterfaceAzure,
+            .container = strDup(container),
+            .account = strDup(account),
+            .key = strDup(key),
+            .blockSize = blockSize,
+            .host = host == NULL ? strNewFmt("%s." AZURE_BLOB_HOST, strPtr(account)) : host,
+            .uriPrefix = host == NULL ?
+                strNewFmt("/%s", strPtr(container)) : strNewFmt("/%s/%s", strPtr(account), strPtr(container)),
+        };
+
+        // Create the http client used to service requests
+        driver->httpClient = httpClientNew(driver->host, port, timeout, verifyPeer, caFile, caPath);
+
+        // Create list of redacted headers
+        driver->headerRedactList = strLstNew();
+        strLstAdd(driver->headerRedactList, HTTP_HEADER_AUTHORIZATION_STR);
+        strLstAdd(driver->headerRedactList, HTTP_HEADER_DATE_STR);
+
+        // Generate starting file id
+        cryptoRandomBytes((unsigned char *)&driver->fileId, sizeof(driver->fileId));
```

Comment:
It's actually the combination of file name and file id that should be unique, but yes. So, this is not perfect, but if two processes are writing against the same file then we already have a problem, currently. This code just tries to ensure each process will write a correct file, even if one overwrites the other, because this is something we may want to do in the future so it could cause some weird bugs if it doesn't work.

On the other hand, the system has a limited amount of random data, so I'd rather not ask for too much and be forced to wait for it. Every process starts at a random place in the 64-bit space and hopefully does not overlap with another process on the same file. The increment is so that a process can also never overlap with *itself*.